### PR TITLE
Add limit for incoming connections from peers without channels

### PR DESCRIPTION
**Pull request by remyers**
Limit the number of incoming connections from nodes we don't have channels with.

- when an incoming connection comes in, the `switchboard` checks if we have channels with that peer
- if we have channels with an incoming peer, do not  track it
- if an incoming peer is on the `sync-whitelist`, do not track it
- otherwise, track the connection 
- when we reach our configured limit of tracked peers without channels, disconnect the oldest tracked peer
- when a tracked peer disconnects or creates a channel, we will stop tracking that peer

We do not track peers we initiate outgoing channel connections to

### CODE REVIEW

**Comment by remyers** [comment]
> We should think about how this interacts with onion messages.

Good point; onion messages are going to be a new source of connections  from nodes without channels.  I didn't see anything in the [Onion Message PR](https://github.com/lightning/bolts/pull/759) for Onion Messages about whether or not a connection should be kept open indefinitely after sending an onion message. This could lead to hitting the limit proposed by this PR and cause an unintentional DoS for both opening channels and sending onion messages. Without any limit it could also lead to an ever increasing number of connections.

As suggested, one approach is to track a timestamp for each connection without a channel and periodically close inactive connections. This would require adding a new data structures in `Switchboard`, or another dedicated class. I was trying to avoid that, but I'll take a look at what would be needed.

**Comment by remyers** [comment]
This PR implements similar functionality to [LDK PR #1988](https://github.com/lightningdevkit/rust-lightning/pull/1988) wrt limiting incoming connections.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+  case class TrackIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private case class ForgetIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private[io] case class CountIncomingConnections(replyTo: ActorRef[Int]) extends Command
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, switchboard: ActorRef[Disconnect]): Behavior[Command] = {
+    Behaviors.setup { context =>
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[PeerDisconnected](c => ForgetIncomingConnection(c.nodeId)))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelCreated](c => ForgetIncomingConnection(c.remoteNodeId)))
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala:41]: The issue is that `ChannelCreated` doesn't mean the channel opening flow will complete, a malicious peer will simply send `open_channel` and then do nothing, and that will let them be forgotten from this rate-limit.

Here is what I think we should do, let me know if you think this is a good idea. We remove this line, and only listen to `PeerDisconnected` events. Once we've reached our rate-limit, we choose a peer that we want to evict (with your current code, the one with the oldest connection) and:

- create a short-lived actor that will query its channel states (a bit like what the `PeerReadyNotifier` does)
- if any of that peer's channels has reached at least the `WAIT_FOR_FUNDING_CONFIRMED` state, which means that there is really a channel and we should keep the connection, we do a `ForgetIncomingConnection` for this peer and start trying to evict the next one in the list with the same process
- if that peer has no channel or none of his channels have made enough progress, we disconnect it

The benefit of that is that it complements nicely the work you've done on limiting the number of pending channels, because that creates a mechanism that will actively close connections with pending channels once we've reached our rate-limit. What do you think about that?

File: eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      new IncomingConnectionsTracker(nodeParams, switchboard).tracking(Map())
+    }
+  }
+}
+
+private class IncomingConnectionsTracker(nodeParams: NodeParams, switchboard: ActorRef[Disconnect]) {
+  import IncomingConnectionsTracker._
+
+  private def tracking(incomingConnections: Map[PublicKey, TimestampMillis]): Behavior[Command] = {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala:50]: It's probably worth a comment on this actor explaining that this doesn't completely solve DoS issues: an attacker could trivially keep opening new connection with us in a loop with random `node_id`s, and that will force us to constantly disconnect every one else (because honest peers won't even have enough time to open a channel before they get disconnected).

This is somewhat ok, because it only affects non-whitelisted incoming connections from peers with whom we didn't have previous channels, so it only limits your ability to receive new incoming liquidity "for free". To work around this, node operators should probably add many known popular nodes to their `sync-whitelist` to make sure they avoid the rate-limits.

I don't think it's this actor's responsibility to do more than what it's already doing: the only way to meaningfully protect against an attacker constantly opening connections with random `node_id`s would be to rate-limit based on IP/AS buckets, which we don't want to do at the application level. This is something that should be done at the TCP level before connections reach eclair, by adding a "standard" DoS-protecting TCP load balancer or something similar.

**Review by t-bast — COMMENTED** [review]
Concept ACK! We'll probably need to beef up the release notes with more details about DoS issues at various level, I believe this should help node operators configure their eclair node to make sure this new feature doesn't create more problems that it solves.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala (hunk 1 above)
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala:41]: I see what you mean. There's  currently no mechanism to terminate incoming channel open requests that do not proceed to the `WAIT_FOR_FUNDING_CONFIRMED` stage. But I believe this should be handled by the channel FSM, not a separate actor.

I feel like it would be easier to reason about this if `IncomingConnectionsTracker` did not need to query the internal state of a channel actor, but only know that there was a hand-off to a channel actor via  `ChannelCreated`. It would then be up to the channel actor to cancel the open request and kill itself if the initiator becomes non-responsive. When the last channel of a `Peer` terminates, the existing logic would kick in that terminates the `Peer` and closes its `PeerConnection`. Once a connection begins the process of creating a channel, it should succeed or close due to a timeout.

An attacker that opens a connection and then requests to open a channel could create free connections that would not count against the `peer-connection.max-no-channels` limit, but it would still be limited by `eclair.channel.maxTotalPendingChannelsPrivateNodes` because `PendingChannelsRateLimiter` limits the number of pending channels that have not completed. An attacker could only freely create connections up to the sum of these limits, not an unlimited number of connections.

Currently an ***initiator*** of channel open requests can set `OpenChannel.timeout_opt` to terminate the process if the counter-party becomes unresponsive - but only up until it reaches a committed state; either `WAIT_FOR_DUAL_FUNDING_CONFIRMED` or `WAIT_FOR_FUNDING_CONFIRMED`.

Why can't we also use this timeout mechanism as a ***non-initiator*** ? Especially in the case of dual-funding, the distinction between being an initiator or not is less clear when both sides may be adding funds.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala (hunk 2 above)
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala:50]: We could do like we do in the `PendingChannelsRateLimiter` and keep two lists: one for public and one for private nodes so that only free liquidity from private nodes could be DoS'd that way. On the other hand, it seems like searching our list of public nodes for each incoming connection could also be a form of resource DoS. I'll just update the comment and we can iterate if an attack materializes.

Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala:54]: Is this too wordy? [3ac3d49](https://github.com/ACINQ/eclair/pull/2601/commits/3ac3d49a416b43d39eca31c8fde49b3650bd1beb)

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala [hunk 3]
```diff
@@ -134,7 +141,7 @@ class Switchboard(nodeParams: NodeParams, peerFactory: Switchboard.PeerFactory)
     getPeer(remoteNodeId) match {
       case Some(peer) => peer
       case None =>
-        log.debug(s"creating new peer (current={})", context.children.size)
+        log.debug(s"creating new peer (current={})", context.children.size - 1)
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala:145]: Fixed in [3ac3d49](https://github.com/ACINQ/eclair/pull/2601/commits/3ac3d49a416b43d39eca31c8fde49b3650bd1beb).

File: eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala (hunk 1 above)
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala]: +1 to not increasing the complexity in the Channel FSM.

I like your idea of listening for `ChannelOpened`! It did not occur to me before that it's not a problem to disconnect a peer while they wait for on-chain confirmation and that the window of time while negotiating should normally be brief with honest peers.  Unless a node is being actively DoS'd (a limitation previously mentioned for this PR) a channel open should always be able to finish negotiation before being closed.

Fixed in [d625549](https://github.com/ACINQ/eclair/pull/2601/commits/d62554966e18aaf2521cb4403d25f87e41f8c292)

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala]: It would be useful to make a couple of manual e2e tests to check how the `ReconnectionTask` interacts with that. I believe that it will make eclair automatically create an _outgoing_ connection (based on a random timer) if we evict a connection where we have a channel waiting for the funding transaction to confirm. That behavior would be a good thing.

Can you test the following scenarios and let me know how eclair behaves?

- Scenario 1
  - Accept an incoming connection which starts opening a channel
  - Simulate a malicious/buggy peer that stops contributing before we reach the state where we're waiting for the transaction to confirm (you'll have to modify eclair to avoid sending `funding_signed` on the fundee side for example)
  - Evict that connection (by having another node connect for example)
  - Verify that eclair does **not** automatically reconnect out to the malicious/buggy peer, even after the `ReconnectionTask` timer
- Scenario 2
  - Accept an incoming connection which starts opening a channel
  - Reach the state where we're waiting for the funding transaction to confirm (both peers are honest)
  - Evict that connection (by having another node connect for example)
  - Verify that eclair automatically reconnects **out** to that peer, which thus bypasses the rate limits (because it's an outgoing connection)

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala [hunk 4]
```diff
@@ -95,15 +98,19 @@ class Switchboard(nodeParams: NodeParams, peerFactory: Switchboard.PeerFactory)
...
+      if (!hasChannels && !authenticated.outgoing) {
+        incomingConnectionsTracker ! TrackIncomingConnection(authenticated.remoteNodeId)
+      }
 
     case ChannelIdAssigned(_, remoteNodeId, _, _) => context.become(normal(peersWithChannels + remoteNodeId))
 
     case LastChannelClosed(_, remoteNodeId) => context.become(normal(peersWithChannels - remoteNodeId))
 
-    case GetPeers => sender() ! context.children
+    case GetPeers => sender() ! context.children.filterNot(_.path.name.startsWith("incoming-connections-tracker"))
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala:113]:
```suggestion
    case GetPeers => sender() ! context.children.filterNot(_ == incomingConnectionsTracker.toClassic)
```

File: eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ * When a tracked peer disconnects or confirms a channel, we will stop tracking that peer.
+ *
+ * We do not need to track peers that disconnect because they will terminate if they have no channels.
+ * Likewise, peers with channels will disconnect and terminate when their last channel closes.
+ *
+ * Note: Peers on the sync whitelist are not tracked.
+
+ * This actor enables a DoS attack that can block new incoming liquidity from non-whitelisted nodes.
+ * An attacker can trivially disconnect other incoming connections before they can open channels by repeatedly opening
+ * new connections using random node IDs.
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala:32]: I think it's worth being more explicit, "DoS attack" is scary (compared to what the attack actually is). It's also useful to mention the (future) fix, what about (also note the missing `*` in the line above):

```suggestion
 *
 * This rate-limiting can be abused by attackers to prevent us from accepting channels from unknown peers: attackers can
 * create a continuous stream of incoming connections with random nodeIds, which forces us to constantly disconnect old
 * connections before they have the opportunity to open a channel. This can be fixed by adding a TCP rate-limiter that
 * rejects connections based on IP addresses, which forces the attacker to own a lot of IP addresses.
```

File: eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  case class TrackIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private case class ForgetIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private[io] case class CountIncomingConnections(replyTo: ActorRef[Int]) extends Command
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, switchboard: ActorRef[Disconnect]): Behavior[Command] = {
+    Behaviors.setup { context =>
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[PeerDisconnected](c => ForgetIncomingConnection(c.nodeId)))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelOpened](c => ForgetIncomingConnection(c.remoteNodeId)))
+      new IncomingConnectionsTracker(nodeParams, switchboard, context).tracking(Map())
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala:47]: very nit: `Map.empty` is more explicit IMO:

```suggestion
      new IncomingConnectionsTracker(nodeParams, switchboard, context).tracking(Map.empty)
```

Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala:47]: It can be useful to tag those logs as being related to connection stuff:

```suggestion
      Behaviors.withMdc(Logs.mdc(category_opt = Some(LogCategory.CONNECTION))) {
        context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[PeerDisconnected](c => ForgetIncomingConnection(c.nodeId)))
        context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelOpened](c => ForgetIncomingConnection(c.remoteNodeId)))
        new IncomingConnectionsTracker(nodeParams, switchboard, context).tracking(Map())
      }
```

File: eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    Behaviors.receiveMessage {
+      case TrackIncomingConnection(remoteNodeId) =>
+        if (nodeParams.syncWhitelist.contains(remoteNodeId)) {
+          Behaviors.same
+        } else {
+          if (incomingConnections.size >= nodeParams.peerConnectionConf.maxNoChannels) {
+            Metrics.IncomingConnectionsDisconnected.withoutTags().increment()
+            val oldest = incomingConnections.minBy(_._2)._1
+            context.log.warn(s"disconnecting peer=$oldest, too many incoming connections from peers without channels.")
+            switchboard ! Disconnect(oldest)
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala:66]: This creates an issue if the `Switchboard` cannot find that `Peer`: the switchboard will send back a `Status.Failure` to the `IncomingConnectionsTracker`, which will crash the actor because it cannot convert it to a `IncomingConnectionsTracker.Command` :cry: 

Maybe we should update `Disconnect` to take a `boolean` argument to indicate whether we expect a response or not? It looks like most of the codebase actually just wants a fire-and-forget message without a response, the only exception is when we call this from the API.

Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala]: Fixed in [be5366a](https://github.com/ACINQ/eclair/pull/2601/commits/be5366a4a35f4c13f0f6499f8ad41e78e37d2fa8). I took the second cleaner fix option, but for consistency, `None` replies to the `sender()` as before and I pass `Some(context.system.ignoreRef[DisconnectResponse])` as the `replyTo_opt` parameter in `IncomingConnectionTracker` to prevent receiving an unhandled reply.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala (hunk 5 above)
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala]: Good point; your suggestion is more clear/correct. Fixed in [9f70d19](https://github.com/ACINQ/eclair/pull/2601/commits/9f70d19d703f72d494daca0171ec6e7e63382737).

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala (hunk 4 above)
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala]: I also added a test to `SwitchboardSpec` to detect if any other other child actor gets created by `Switchboard` and returned by `GetPeers`.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 8]
```diff
@@ -523,5 +541,4 @@ object Peer {
   case class RelayOnionMessage(messageId: ByteVector32, msg: OnionMessage, replyTo_opt: Option[typed.ActorRef[Status]])
 
   case class RelayUnknownMessage(unknownMessage: UnknownMessage)
-  // @formatter:on
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:526]: nit: why did you remove this? When we create a `formatter:off` block, we always end it with a `formatter:on`

File: eclair-core/src/test/scala/fr/acinq/eclair/io/SwitchboardSpec.scala [hunk 9]
```diff
@@ -141,6 +144,66 @@ class SwitchboardSpec extends TestKitBaseClass with AnyFunSuiteLike {
...
+    switchboard ! PeerConnection.Authenticated(peerConnection.ref, unknownNodeId1, outgoing = false)
+    peer.expectMsgType[Peer.Init]
+    switchboard ! PeerConnection.Authenticated(peerConnection.ref, unknownNodeId2, outgoing = false)
+    peer.expectMsgType[Peer.Init]
+
+    // Disconnect the oldest tracked peer when an incoming connection from a peer without channels connects.
+    switchboard ! PeerConnection.Authenticated(peerConnection.ref, randomKey().publicKey, outgoing = false)
+    peer.fishForMessage() {
+      case d: Peer.Disconnect => d.nodeId == unknownNodeId1
+      case _: Peer.Init => false
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/io/SwitchboardSpec.scala:176]: This case is unnecessary, `fishForMessage()` take a `PartialFunction`. You can also remove it line 189 below.

File: eclair-core/src/test/scala/fr/acinq/eclair/io/SwitchboardSpec.scala [hunk 10]
```diff
@@ -141,6 +144,66 @@ class SwitchboardSpec extends TestKitBaseClass with AnyFunSuiteLike {
...
+
+    // Do not disconnect an old peer when a peer with channels connects.
+    switchboard ! ChannelIdAssigned(channel.ref, hasChannelsNodeId2, randomBytes32(), randomBytes32())
+    switchboard ! PeerConnection.Authenticated(peerConnection.ref, hasChannelsNodeId2, outgoing = false)
+    peer.expectMsgType[Peer.Init]
+    peer.expectNoMessage(100 millis)
+
+    // Disconnect the next oldest tracked peer when an incoming connection from a peer without channels connects.
+    switchboard ! PeerConnection.Authenticated(peerConnection.ref, randomKey().publicKey, outgoing = false)
+    peer.fishForMessage()  {
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/io/SwitchboardSpec.scala:187]: nit:
```suggestion
    peer.fishForMessage() {
```

File: eclair-core/src/test/scala/fr/acinq/eclair/io/IncomingConnectionsTrackerSpec.scala [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  test("stop tracking a node that disconnects and free space for a new node connection") { f =>
+    import f._
+
+    // Track nodes without channels.
+    tracker ! IncomingConnectionsTracker.TrackIncomingConnection(connection1)
+    tracker ! IncomingConnectionsTracker.TrackIncomingConnection(connection2)
+    monitorProbe.expectMessageType[TrackIncomingConnection]
+    monitorProbe.expectMessageType[TrackIncomingConnection]
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/io/IncomingConnectionsTrackerSpec.scala:68]: I don't understand the goal of these assertions, they aren't testing anything, are they?
If you want to make sure that the two messages above have been received, you should rather do:

```scala
    eventually {
      tracker ! IncomingConnectionsTracker.CountIncomingConnections(probe.ref)
      probe.expectMessage(2)
    }
```

Same elsewhere, the `monitorProbe` assertions seem useless...I believe you initially used it to test that the actor crashed if sent `Status.Failure`, but you can probably remove `monitorProbe` entirely?

File: eclair-core/src/test/scala/fr/acinq/eclair/io/SwitchboardSpec.scala (hunk 9 above)
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/SwitchboardSpec.scala:176]: I use `fishForMessage` because usually `Init` is sent to `peer` but sometimes it can happen that `Peer.Disconnect` is seen first. I'm using it to ignore `Init` and wait for `Peer.Disconnect`. If I remove that line, I get the following error. 
```
assertion failed: fishForMessage() found unexpected message Init(Set())
```

Comment (t-bast) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/SwitchboardSpec.scala:176]: Damn, that means it doesn't really accept just a partial function...but then you should ignore *all* other messages, not just `Peer.Init`? It should be `case _ => false`?

File: eclair-core/src/test/scala/fr/acinq/eclair/io/IncomingConnectionsTrackerSpec.scala (hunk 11 above)
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/IncomingConnectionsTrackerSpec.scala]: I forgot now why I reintroduced that; I think I was looking for a way to detect when the tracker dies due to the original issue you found about receiving an unhandled message from `switchboard`. Anyway, it doesn't need a test with the current fix because now `Peer.Disconnect` never has a reference to the `incomingConnectionsTracker`. Removed in [a00f448](https://github.com/ACINQ/eclair/pull/2601/commits/a00f448df3d45b3ab26952948a3c47f6cfeec9c0).

File: eclair-core/src/test/scala/fr/acinq/eclair/io/SwitchboardSpec.scala (hunk 10 above)
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/SwitchboardSpec.scala]: fixed in [a00f448](https://github.com/ACINQ/eclair/pull/2601/commits/a00f448df3d45b3ab26952948a3c47f6cfeec9c0)

File: eclair-core/src/test/scala/fr/acinq/eclair/EclairImplSpec.scala [hunk 12]
```diff
@@ -697,4 +697,15 @@ class EclairImplSpec extends TestKitBaseClass with FixtureAnyFunSuiteLike with I
...
+    import f._
+
+    val eclair = new EclairImpl(kit)
+
+    eclair.disconnect(randomKey().publicKey).pipeTo(sender.ref)
+    val disconnect = switchboard.expectMsgType[Disconnect]
+    val replyTo = disconnect.replyTo_opt.getOrElse(sender.ref.toTyped)
+    replyTo ! Disconnecting(disconnect.nodeId)
+    sender.expectMsg(Disconnecting(disconnect.nodeId))
+  }
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/EclairImplSpec.scala]: removed in [a00f448](https://github.com/ACINQ/eclair/pull/2601/commits/a00f448df3d45b3ab26952948a3c47f6cfeec9c0)

File: eclair-core/src/main/scala/fr/acinq/eclair/io/PeerConnection.scala [hunk 13]
```diff
@@ -464,6 +464,9 @@ class PeerConnection(keyPair: KeyPair, conf: PeerConnection.Conf, switchboard: A
     case StopEvent(_, CONNECTED, d: ConnectedData) =>
       Metrics.PeerConnectionsConnected.withoutTags().decrement()
       d.peer ! Peer.ConnectionDown(self)
+    case StopEvent(_, INITIALIZING, d: InitializingData) =>
+      log.debug(s"terminated while initializing.")
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/PeerConnection.scala]: fixed in [a00f448](https://github.com/ACINQ/eclair/pull/2601/commits/a00f448df3d45b3ab26952948a3c47f6cfeec9c0)

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala (hunk 8 above)
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala]: I'm not sure how I missed that, fixed in [a00f448](https://github.com/ACINQ/eclair/pull/2601/commits/a00f448df3d45b3ab26952948a3c47f6cfeec9c0)

**Review by t-bast — APPROVED** [review]
LGTM :rocket:
